feat: Blogpost: NuttX motor control and sensing part 1/2 (#480)
espressif/developer-portal: content/blog/nuttx-motor-control-and-sensing/index.md

---
title: "NuttX for Motor Control and Sensing 1/2"
date: 2025-04-28

> tags: ["NuttX", "Apache", "ESP32", "Motor", "IMU"]

Reviewer: Should you add ESP32-C6 in the tag list?

showAuthor: false
authors:
    - "filipe-cavalcanti"
summary: "This article demonstrates how to implement motor control and speed sensing on an ESP32-C6 using NuttX RTOS. It covers setting up MCPWM for motor control, ADC for potentiometer reading, and quadrature encoder for speed measurement. The implementation showcases NuttX's real-time capabilities through a practical example using common peripherals and sensors."
---

## Introduction

> This series of articles demonstrates how to use multiple ESP32-C6 peripherals to build a complete motor control and sensing system. We'll explore:

Reviewer: I would suggest avoiding the use of "We" in overall documentation. @pedrominatel any recommendation about it or it is good as is?

Author: I would say this applies to academic articles. Many articles on the portal are "friendly".

- Motor Control PWM (MCPWM) for DC motor speed control
- Quadrature encoder (pulse counter) for motor speed measurement
- ADC for real-time potentiometer position reading
- Inertial Measurement Unit (IMU) via I2C for vibration analysis
- Wireless connectivity for data acquisition and processing

> The goal is to build a comprehensive motor control and sensing system using the NuttX RTOS. We'll implement open-loop speed control while measuring mechanical vibrations through an IMU. The system will stream sensor data wirelessly to a computer for logging and analysis, showcasing NuttX's real-time capabilities, multi-threading support, and the ESP32-C6's wireless connectivity features.

Reviewer: Remove last comma: `multi-threading support and the ESP32-C6's wireless connectivity features.`

> In this first article, we'll focus on implementing basic motor control. You'll learn how to use the ESP32-C6's MCPWM peripheral to drive a DC motor through an H-bridge, with speed control provided by a potentiometer input and reading the quadrature encoder. We'll implement this using the NuttX RTOS, preparing the basics before we can get to data offloading and analysis in the next article.

Reviewer: I sugest removing `We'll implement this using the NuttX RTOS` as it is already known that NuttX will be used :)

As the starting point for our project, we create an empty repository that will be used as an extension of `nuttx-apps`. This approach allows simple integration with the NuttX build system, just like a standard application.

### Hardware Requirements

The following components are used in this project:

- CHR-GM25-370 6V DC Motor with integrated gearbox and quadrature encoder
- L298N H-Bridge motor driver module
- 10kΩ linear potentiometer
- GY521 IMU module (MPU6050)
- ESP32C6-DevkitC development board

All components have been mounted on a test bench as shown below:

{{< figure
    default=true
    src="img/test_bench.png"
    >}}

## Motor Setup
For motor control using the H-Bridge, we can use either one or two pins for PWM signals. Using one pin works but the motor will only spin in one direction, while two pins allows us to control both direction (forward and reverse) and speed.

While NuttX and the Espressif MCPWM driver support both modes, this application focuses on unidirectional control using a single GPIO. We'll use PWM_A on GPIO 20 of the ESP32-C6 to control motor speed.

> Our setup uses an L298N H-Bridge module driving a 6V DC motor. The motor includes a gearbox that reduces the output speed to 210 RPM. A quadrature encoder attached to the gearbox shaft will be used for speed measurement. The encoder is powered on the 3.3 V bus and channels 1 and 2 of the encoder are connected to GPIO pins 10 and 11.

Reviewer: Sounds better making it `3.3V`

## Potentiometer Setup
Motor speed control is implemented using a normalized range from 0.0 to 1.0. A potentiometer provides manual control over this range, allowing intuitive speed adjustment.

> The ESP32's built-in ADC support in NuttX makes this implementation straightforward. We'll connect the potentiometer's wiper to GPIO 3 (ADC1 Channel 3) and power it with the board's 3.3V supply.

Reviewer: Not necessary to change, but you could avoid those apostrophes in 'ESP32` or `ESP32-C6` around the documentation. Ex.:
`Built-in ADC support in the ESP32 makes this implementation straightforward.`.

Author: Agreed

In menuconfig, navigate to `System Type → Peripheral Support` and enable `Pulse Counter (PCNT / QE) Module`. This will open a configuration menu where we need to set:

- `Enable PCNT Unit 0`
- `Use this PCNT Unit as Quadrature Encoder`
- `Enable Glitch Filter for this PCNT Unit`

The quadrature encoder driver operates in X4 mode by default, providing maximum resolution for position and speed measurement. Configure the GPIO pins as follows:

```
PCNT Unit 0 Configuration:
- CH0 Edge/Pulse Pin:   GPIO 10
- CH0 Level/Control:    GPIO 11
- CH1 Edge/Pulse Pin:   GPIO 10
- CH1 Level/Control:    GPIO 11
```

Finally, enable the generic quadrature encoder driver support:
1. Navigate to `Device Drivers`
2. Enable `Sensor Device Support`
3. Enter the Sensor menu and enable `Qencoder`

{{< figure
    default=true
    src="img/menuconfig_qencoder.png"
    >}}

After configuration, the quadrature encoder will be available as a character device at `/dev/qe0`.

### Setup ADC

> The ESP32-C6 features a 12-bit SAR ADC that we'll use for reading the potentiometer position. Although NuttX provides a ready-to-use `adc` defconfig, following the manual configuration steps helps understand how the ADC peripheral is set up for our specific application.

Reviewer: ```suggestion
With its 12-bit SAR ADC, the ESP32-C6 can directly read the potentiometer position. NuttX provides an out-of-the-box adc defconfig, but going through the manual setup steps reveals how the ADC driver is actually wired up under the hood.
```

In menuconfig, navigate to `System Type → Peripheral Support` and enable the ADC driver. The ESP32-C6 provides one ADC unit, while some other ESP32 variants offer two units.

{{< figure
    default=true
    src="img/menuconfig_adc_1.png"
    >}}

Under `System Type → ADC Configuration`, you can configure the ADC's input attenuation, operation mode, and channel selection. Navigate to `ADC1 Channel Selection` and enable ADC Channel 3 (GPIO 3).

{{< figure
    default=true
    src="img/menuconfig_adc_2.png"
    >}}

Each ADC channel maps to a specific GPIO pin. You can find these mappings in the ESP32-C6 datasheet or the NuttX [documentation for ADC](https://nuttx.apache.org/docs/latest/platforms/risc-v/esp32c6/index.html#analog-to-digital-converter-adc).

For debugging purposes, it's recommended to enable the ADC example application. Navigate to `Application Configuration → Examples` and enable `ADC Example` along with `Use software trigger`.

### Verify Driver Registration

> After configuring the drivers, let's verify they're properly registered in the system. Build and flash the firmware:

Reviewer: ```suggestion
Once the drivers are configured, verify they’ve been registered correctly in the system. Then build and flash the firmware:
```

- Emergency motor stop capability
- ADC sampling from the potentiometer
- Real-time conversion of ADC readings to motor speed values
- Continuous speed updates based on potentiometer position
- Shows speed measurement in RPM from the encoder data

### Code Structure

The application follows this directory structure:

```
.
├── apps
│   ├── Make.defs
│   ├── Makefile
│   ├── motor_sensing
│   │   ├── config
│   │   │   └── defconfig
│   │   ├── Kconfig
│   │   ├── Make.defs
│   │   ├── Makefile
│   │   └── motor_sensing_main.c
├── LICENSE
└── README.md
```

The main implementation resides in `motor_sensing_main.c`, while our saved configuration is stored in `apps/motor_sensing/config/defconfig`.

For details on external application integration with NuttX, refer to the [official documentation](https://nuttx.apache.org/docs/latest/guides/building_nuttx_with_app_out_of_src_tree.html).

This entire application is available on a [GitHub repository](https://github.com/fdcavalcanti/nuttx-esp-motor-sensing).

### Implementation Details

First, we define the constants needed for our application. These include device paths, ADC thresholds, and encoder parameters:

```C

> #define MOTOR_DEVPATH "/dev/motor0"

Author: Did manually

#define ADC_DEVPATH "/dev/adc0"
#define QE_DEVPATH "/dev/qe0"
#define ADC_MIN_THRESHOLD 100
#define ADC_MAX_THRESHOLD 2500
#define TASK_DELAY_MS 100
#define BASE_PPR 11        /* Base encoder PPR */
#define GEAR_RATIO 34      /* Gear reduction ratio */
#define PULSES_PER_REV (BASE_PPR * GEAR_RATIO)

> ```

Reviewer: By looking at the review page here, sounds the formatting isn't ok here.. PTAL.

Author: Seems the Markdown does not color the defines.

I added some functions and it highlighted fine on the same code block.

  if (sample.am_data < ADC_MIN_THRESHOLD)
    {
      *speed = 0.0;
    }
  else if (sample.am_data > ADC_MAX_THRESHOLD)
    {
      *speed = 1.0;
    }
  else
    {
      /* Linear mapping from ADC range to speed range */
      *speed = (float)(sample.am_data - ADC_MIN_THRESHOLD) /
               (float)(ADC_MAX_THRESHOLD - ADC_MIN_THRESHOLD);
    }

  printf("ADC Value: %" PRId32 " | Motor Speed: %.2f\n", sample.am_data, *speed);
  return OK;
}
```

Regarding quadrature encoder, by consulting `include/nuttx/sensors/qencoder.h`, we can access the quadrature encoder API. Our control loop should read the quadrature encoder position each iteration, reset it and calculate the motor speed in RPM. The speed calculation takes into account:
- The X4 encoding mode, which generates 4 counts per encoder pulse
- The gearbox reduction ratio
- The base encoder resolution (pulses per revolution)
- The time elapsed between measurements

```C
static float calculate_rpm(int32_t pulses, uint32_t time_ms)
{
  /* Convert encoder pulses to RPM:
   * RPM = (pulses/4 / PULSES_PER_REV) * (60000 / time_ms)
   * Note: divide by 4 because driver uses X4 encoding by default
   */

>   return ((float)(pulses/4) * 60000.0f) / ((float)PULSES_PER_REV * (float)time_ms);

Reviewer: You are probably missing some rpm range due to integer calculation at `pulses/4`.
```suggestion
  return (float)pulses * 60000.0f / (4.0f * (float)PULSES_PER_REV * (float)time_ms);
```

Author: You are right, it was off by some 4 ou 5 RPM.
Thanks!
